Fix default timeouts for python entrypoints (e.g. init_process_group)

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113044
* __->__ #112893
* #112803
* #112611

Previous PRs changed the c++ default timeout for PGNccl, but this path
was only hit in some cases, and the python defaults took over in other
cases.

This PR ensures that NCCL pg always default to the changed NCCL-specific
timeout value.

wconstab committed:
Fix default timeouts for python entrypoints (e.g. init_process_group)

Previous PRs changed the c++ default timeout for PGNccl, but this path
was only hit in some cases, and the python defaults took over in other
cases.

This PR ensures that NCCL pg always default to the changed NCCL-specific
timeout value.

[ghstack-poisoned]

xw285cornell reviewed (approved):
LGTM!

kwen2501 reviewed (approved):
LGTM. 
Hopefully we can make the unification later.
Maybe we should also update doc of the APIs: `timeout=datetime.timedelta(seconds=1800)`

wconstab committed:
Update on "Fix default timeouts for python entrypoints (e.g. init_process_group)"


Previous PRs changed the c++ default timeout for PGNccl, but this path
was only hit in some cases, and the python defaults took over in other
cases.

This PR ensures that NCCL pg always default to the changed NCCL-specific
timeout value.

[ghstack-poisoned]

XilunWu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1225,10 +1238,8 @@ def _new_process_group_helper(
-            "Expected timeout argument to be of type datetime.timedelta"
-        )
+    # Note: _new_process_group_helper is only called from init_process_group, which always provides a timeout value
+    _check_valid_timeout(timeout)
```
isn't the `timeout` argument passed in already checked in `init_process_group`? My thought is we don't need this check in all private APIs (i.e. `_new_process_group_helper` and `_new_group_with_tag`) because they must have been checked in the public API before being passed in.

wconstab replied:
init_process_group checks the timeout and then calls _new_process_group_helper

other 'new_group' apis also call _new_process_Group_helper directly.

My first thought was to move all the checking inside the helper and take it out of all the outer functions.  I still think this is a good idea.  However, this is a larger change since init_process_group also has handling for `pg_options` timeout value.  Should I fix this now or in a separate pr...?

XilunWu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -3927,11 +3913,19 @@ def _new_group_with_tag(
         backend = default_backend
     backend = Backend(backend)
+
+    # this timeout defaulting/validation is used for all the new_groups/new_subgroups variants,
```
I don't think we can reduce the appearance of this defaulting/validation which is sad... Any idea?

XilunWu reviewed (approved):
LGTM. Left some thoughts

wconstab committed:
Update on "Fix default timeouts for python entrypoints (e.g. init_process_group)"


Previous PRs changed the c++ default timeout for PGNccl, but this path
was only hit in some cases, and the python defaults took over in other
cases.

This PR ensures that NCCL pg always default to the changed NCCL-specific
timeout value.

[ghstack-poisoned]

wconstab committed:
Update on "Fix default timeouts for python entrypoints (e.g. init_process_group)"


Previous PRs changed the c++ default timeout for PGNccl, but this path
was only hit in some cases, and the python defaults took over in other
cases.

This PR ensures that NCCL pg always default to the changed NCCL-specific
timeout value.

[ghstack-poisoned]

kwen2501 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1060,26 +1074,14 @@ def init_process_group(
-            async error handling is done differently since with UCC we have
-            progress thread and not watch-dog thread.
+            the process group. Default value is 10 minutes for NCCL and 30 minutes for other backends.
+            This is the duration after which collectives will be aborted asynchronously and the process will crash.
+            When ``NCCL_ASYNC_ERROR_HANDLING`` is set, this is the duration after which collectives will be aborted
+            asynchronously and the process will crash.
```
Duplicated?

kwen2501 reviewed (approved):
The updates look good to me

wconstab commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "sorry this seems to have broken inductor https://hud.pytorch.org/pytorch/pytorch/commit/f9d47e13813bbefc9f19a6c0430b7122f9d09b91 https://github.com/pytorch/pytorch/actions/runs/6776367936/job/18418174752" -c nosignal

```

2023-11-06T21:45:40.8453227Z 
2023-11-06T21:45:41.1327996Z loading model: 0it [00:00, ?it/s]
2023-11-06T21:45:41.1328806Z loading model: 0it [00:00, ?it/s]
2023-11-06T21:45:41.1337717Z Traceback (most recent call last):
2023-11-06T21:45:41.1338781Z   File "benchmarks/dynamo/torchbench.py", line 532, in <module>
2023-11-06T21:45:41.1339770Z     torchbench_main()
2023-11-06T21:45:41.1340677Z   File "benchmarks/dynamo/torchbench.py", line 528, in torchbench_main
2023-11-06T21:45:41.1341834Z     main(TorchBenchmarkRunner(), original_dir)
2023-11-06T21:45:41.1343184Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3362, in main
2023-11-06T21:45:41.1347414Z     process_entry(0, runner, original_dir, args)
2023-11-06T21:45:41.1348900Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3319, in process_entry
2023-11-06T21:45:41.1353205Z     return maybe_fresh_cache(
2023-11-06T21:45:41.1354422Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 1939, in inner
2023-11-06T21:45:41.1356840Z     return fn(*args, **kwargs)
2023-11-06T21:45:41.1358017Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3801, in run
2023-11-06T21:45:41.1363384Z     ) = runner.load_model(
2023-11-06T21:45:41.1364244Z   File "benchmarks/dynamo/torchbench.py", line 412, in load_model
2023-11-06T21:45:41.1364832Z     benchmark = benchmark_cls(
2023-11-06T21:45:41.1365566Z   File "/var/lib/jenkins/workspace/torchbench/torchbenchmark/util/model.py", line 23, in __call__
2023-11-06T21:45:41.1366375Z     obj = type.__call__(cls, *args, **kwargs)
2023-11-06T21:45:41.1367242Z   File "/var/lib/jenkins/workspace/torchbench/torchbenchmark/models/moco/__init__.py", line 56, in __init__
2023-11-06T21:45:41.1368465Z     dist.init_process_group(backend='nccl', init_method='tcp://localhost:10001',
2023-11-06T21:45:41.1370095Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/c10d_logger.py", line 86, in wrapper
2023-11-06T21:45:41.1370956Z     func_return = func(*args, **kwargs)
2023-11-06T21:45:41.1372010Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1146, in init_process_group
2023-11-06T21:45:41.1383161Z     timeout = _get_default_timeout(backend)
2023-11-06T21:45:41.1385072Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 604, in _get_default_timeout
2023-11-06T21:45:41.1387332Z     assert isinstance(default_pg_nccl_timeout, timedelta), "no NCCL default timeout, is NCCL support compiled?"
2023-11-06T21:45:41.1388842Z AssertionError: no NCCL default timeout, is NCCL support compiled?
2023-11-06T21:45:41.5615499Z ERROR
2023-11-06T21:45:43.8661748Z 
2023-11-06T21:45:46.3864865Z loading model: 0it [00:00, ?it/s]
2023-11-06T21:45:46.3865349Z loading model: 0it [00:02, ?it/s]
```
